### PR TITLE
[Yaml] Fix wrong line break handling at block end for folded style

### PR DESCRIPTION
**Pull request by TiMESPLiNTER**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | yes |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #19670 |
| License | MIT |

I guess there's a BC break because YAML parsing changes.

### CODE REVIEW

**Comment by fabpot** [comment]
Fixing a bug should always be done in the oldest and still maintained version (so 2.7), but BC breaks cannot. So, in this case, I see two possibilities: we fix the bug in 2.7, or we deprecate the current behavior in master and fix the issue in 4.0 (like we've done in the past for similar situations).

**Comment by TiMESPLiNTER** [comment]
@fabpot I'm for the second one (introduce in 4.0). The fix definitely can change the result of the YAML parser for an already existing YAML file. And the code using that value can break if it's not written flexible enough (and I'm pretty sure there's many code out there which this may be the case).

**Comment by nicolas-grekas** [comment]
@TiMESPLiNTER then can you please update your PR so that it triggers a deprecation notice in 3.x?

**Comment by TiMESPLiNTER** [comment]
@nicolas-grekas where and about what should this deprecation notice be? I mean there's not feature depricated but the outcome of the YAML parser would be a different one.

**Comment by nicolas-grekas** [comment]
The way we deal with this is still allow the current behavior, but warn the user that this is valid yaml and won't be supported anymore in 4.0. Is that possible in this case?

**Comment by stof** [comment]
@nicolas-grekas but there is nothing to change in the YAML file for this bug. The issue is not that the parser was accepting invalid YAML (like things we deprecated previously in the parser), but that the parser returns the wrong value when parsing the YAML.

**Comment by nicolas-grekas** [comment]
So, there is no other way than applying this on 2.7 then, isn't it?

**Comment by TiMESPLiNTER** [comment]
@nicolas-grekas but we have to be aware that this may break third party code which is not flexible enough to handle new lines at the end of the string.

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by TiMESPLiNTER** [comment]
@nicolas-grekas so I open up the PR against the 2.7 branch?

**Comment by nicolas-grekas** [comment]
@TiMESPLiNTER no need, we can do it while merging (and you can also do it since github added this possiblity recently in the UI, see "change base branch" feature)

**Comment by TiMESPLiNTER** [comment]
@nicolas-grekas well but I've to implement the fix again on the 2.7 branch because my fix is now based on the current master so if we would change the branch from `master` to `2.7` all master commits would be also part of that PR.

**Comment by nicolas-grekas** [comment]
Then you'd need to rebase on 2.7 right!

**Comment by TiMESPLiNTER** [comment]
@nicolas-grekas okay I switched to 2.7 as base branch. Can I somehow retest because the tests failed because of the rebase.

**Comment by HeahDude** [comment]
Change the hash of your commit (you can do it by adding a dummy commit to be reverted and by squashing everything), then do a forced push, it will reset the test suite.

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 1]
```diff
@@ -1012,7 +1012,7 @@ public function getCommentLikeStringInScalarBlockData()
     public function testBlankLinesAreParsedAsNewLinesInFoldedBlocks()
     {
         $yaml = <<<'EOT'
-test: >
+test: >-
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: I think I would rather change the expected outcome instead.

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 2]
```diff
@@ -1036,7 +1036,7 @@ public function testBlankLinesAreParsedAsNewLinesInFoldedBlocks()
     public function testAdditionallyIndentedLinesAreParsedAsNewLinesInFoldedBlocks()
     {
         $yaml = <<<'EOT'
-test: >
+test: >-
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: same here

File: src/Symfony/Component/Yaml/Parser.php [hunk 3]
```diff
@@ -639,7 +640,7 @@ private function parseBlockScalar($style, $chomping = '', $indentation = 0)
 
         // deal with trailing newlines
         if ('' === $chomping) {
-            $text = preg_replace('/\n+$/', "\n", $text);
+            $text = preg_replace('/\n+$/', "\n", $text.('>' === $style ? "\n" : null));
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: I would append the empty string instead of `null`.

**Comment by xabbuh** [comment]
The change is valid in regards to the YAML spec. However, as this may indeed break existing applications that relied on the existing behaviour of the parser I am inclined to merge this into `master` instead and clearly document the changed behaviour in the changelog and upgrade files.

@symfony/deciders What do you think?

**Comment by javiereguiluz** [comment]
I think this should go to 2.7. We must comply with the Yaml spec and we did not, so this is a bug that must be fixed and not just a BC break.

**Comment by fabpot** [comment]
I would so it on master.

**Comment by fabpot** [comment]
@xabbuh Can you take care of merging this to master then?

**Comment by xabbuh** [comment]
While trying to finish this for `master` I noticed that the proposed fix is not correct either (it would now always add a trailing newline even if the block was at the end of the YAML string to be parsed and didn't contain a trailing newline). I'll look into it and try to see how we can fix it properly.

**Comment by xabbuh** [comment]
I now managed to look into this again. In fact, I cannot reproduce the behaviour described in #19670. And merging this patch would be even wrong as the fix would not only preserve a trailing newline in a folded style block, but would also add one if there wasn't any before (e.g. when EOF is reached). Thus, I am going to close here as I do not see anything to be fixed.

**Pull request closed without merging**